server/epdfinfo.c

```diff
@@ -391,46 +492,107 @@ image_recolor (cairo_surface_t * surface, const PopplerColor * fg,
   for (y = 0; y < page_height * rowstride; y += rowstride)
     {
       unsigned char *data = image + y;
 
       unsigned int x;
       for (x = 0; x < page_width; x++, data += 4)
-	{
-	  /* Careful. data color components blue, green, red. */
-	  const double rgb[3] = {
-	    (double) data[2] / 256.,
-	    (double) data[1] / 256.,
-	    (double) data[0] / 256.
-	  };
-
-	  /* compute h, s, l data   */
-	  double l = a[0] * rgb[0] + a[1] * rgb[1] + a[2] * rgb[2];
-
-	  /* linear interpolation between dark and light with color ligtness as
-	   * a parameter */
-	  data[2] =
-	    (unsigned char) round (255. * (l * rgb_diff[0] + rgb_fg[0]));
-	  data[1] =
-	    (unsigned char) round (255. * (l * rgb_diff[1] + rgb_fg[1]));
-	  data[0] =
-	    (unsigned char) round (255. * (l * rgb_diff[2] + rgb_fg[2]));
-	}
+        {
+          /* Careful. data color components blue, green, red. */
+          struct color rgb = {
+            .r = (double) data[2] / 256.,
+            .g = (double) data[1] / 256.,
+            .b = (double) data[0] / 256.
+          };
+
+          switch (usecolors)
+            {
+            case 0:
+              /* No image recoloring requested.  Do nothing in this case.
+                 Should never be called as we should never call with unless
+                 usecolors != 0. */
+              break;
+            case 1:
+              {
+                /* Linear interpolation between bg and fg based on the
+                   perceptual lightness measure l */
+                /* compute h, s, l data   */
+                double l = a[0] * rgb.r + a[1] * rgb.g + a[2] * rgb.b;
+
+                /* linear interpolation between dark and light with color
+                   lightness as a parameter */
+                data[2] =
+                  (unsigned char) round (255. * (l * rgb_diff.r + rgb_fg.r));
+                data[1] =
+                  (unsigned char) round (255. * (l * rgb_diff.g + rgb_fg.g));
+                data[0] =
+                  (unsigned char) round (255. * (l * rgb_diff.b + rgb_fg.b));
+              }
+              break;
+            case 2:
+              {
+                /* Convert to Oklab coordinates, invert perceived lightness,
+                   convert back to RGB. */
+                if (color_equal(white, rgb))
+                  {
+                    rgb = black;
+                  }
+                else if (color_equal(precomputed_rgb, rgb))
+                  {
+                    rgb = precomputed_inv_rgb;
+                  }
+                else
+                  {
+                    struct color oklab = rgb2oklab(rgb);
+                    precomputed_rgb = rgb;
+
+                    /* Gamma correction.  Shouldn't be necessary, but colors
+                     * 'feel' too dark and fonts too thin otherwise. */
+                    oklab.l = pow(oklab.l, 1.8);
+
+                    /* Invert the perceived lightness */
+                    oklab.l = 1.0 - oklab.l;
```

Comment:
I think we can add a line to make sure `oklab.l` is between the l value of `fg` and `bg` here, what do you think? And add in documentations that `pdf-view-midnight-colors` will be used to determine the lightness in this case, but those colors will be used as it is in the other case. This should help with not letting the contrast be too high, but I'm not sure about the ramification of that. I thought about changing the precomputed white and black that way, but that might not work with colors that are close to them, so I think this implementation should be better.